Backport 8259922: MethodHandles.collectArguments does not throw IAE

babsingh opened the pull request:
Fix for https://bugs.openjdk.java.net/browse/JDK-8259922.

Fix: https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/ab6b9c35c7541ea628b028cb1357b904a72de5d9.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

babsingh commented:
fyi @pshipton Identical to https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/29. For supporting OJDK MHs.

pshipton commented:
https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/29/commits/f6e2b25af55d44 also added a test MethodHandlesCollectArgsTest.java, I think we want that as well?

pshipton commented:
Maybe we'd have to exclude the test until we switch?

pshipton commented:
Why do we need to backport this if OpenJDK hasn't bothered to backport to 8, 11?

babsingh commented:
> Why do we need to backport this if OpenJDK hasn't bothered to backport to 8, 11?

Because there are tests in OpenJ9, which will fail without this fix once OJDK MHs are enabled: https://github.com/eclipse-openj9/openj9/issues/11922.

> We're backporting an OpenJDK change, so we should backport the Oracle copyright, not add an IBM copyright.

Updated Oracle copyright and removed IBM copyright.

> https://github.com/ibmruntimes/openj9-openjdk-jdk16/commit/f6e2b25af55d44 also added a test MethodHandlesCollectArgsTest.java, I think we want that as well?

This test won't work with JDK8 because `MethodHandles.empty` is not available in JDK8. We can either update the test or rely on the existing OJ9 tests for coverage in JDK8. For JDK11, I have included the test in https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/560. 
 
> Maybe we'd have to exclude the test until we switch?

We do not need to exclude the test because the test should correctly work with OJ9 MHs

pshipton commented:
jenkins test sanity,extended alinux jdk8

pshipton commented on jdk/src/share/classes/java/lang/invoke/MethodHandles.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2008, 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2008, 2022, Oracle and/or its affiliates. All rights reserved.
```
https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/ab6b9c35c7541ea628b028cb1357b904a72de5d9 has a 2021 end date.

babsingh replied:
Updated to match the 2021 end date.

pshipton commented:
Oops, PR testing isn't useful here since the class being modified isn't used atm. Luckily it failed anyway because I got the platform name wrong.
